### PR TITLE
HTLC creation and integration with issuance flow

### PR DESCRIPTION
**Pull request by judywu-da**
This PR includes below changes

- add HTLC settlement contracts (Instruction, Delivery in HTLC module)
- integrate HTLC settlement contracts with Tranche Template to enable HTLC settlement support in Issuance Flow
- Code refactoring in InstructSettlement choices in Tranche template

### CODE REVIEW

File: daml/Tests/Distribution/Syndication/Issuance.daml [hunk 1]
```diff
@@ -92,28 +94,42 @@ issuance Parties{..} Origination{..} = do
...
-  settleTrades bondRegistrar [ dvpCid11, dvpCid12, dvpCid13 ]
+  signInstructionsHtlc [ custodian2, settlementBank ]
+  settleTrades bondRegistrar [ dvpCid11, dvpCid12 ]
+  settleTradeHtlc settlementBank (secret, dvpCid13)
 
   -- Step 2
   dvpCid21 <- submit operator do exerciseCmd tranche1Cid Structuring.InstructBndSettlement with price; dateOfSettlement; dateOfTrade
   dvpCid22 <- submit operator do exerciseCmd tranche2Cid Structuring.InstructBndSettlement with price; dateOfSettlement; dateOfTrade
-  dvpCid23 <- submit operator do exerciseCmd tranche3Cid Structuring.InstructBndSettlement with price; dateOfSettlement; dateOfTrade
+  dvpCid23 <- submit operator do exerciseCmd tranche3Cid Structuring.InstructBndSettlementHtlc with price; dateOfSettlement; dateOfTrade; hashlock; expiry
```
Comment (GeorgSchneider) [inline daml/Tests/Distribution/Syndication/Issuance.daml:115]: Should use a different secret / hashlock here - the two settlements are independent

File: daml/Tests/Distribution/Syndication/Issuance.daml [hunk 2]
```diff
@@ -92,28 +94,42 @@ issuance Parties{..} Origination{..} = do
...
+  signInstructionsHtlc [ custodian2, custodian3, bndBank ]
+  settleTrades bondRegistrar [ dvpCid21, dvpCid22 ]
+  settleTradeHtlc bndBank (secret, dvpCid23)
 
-  -- -- Step 3
+  -- -- -- Step 3
   dvpCids31 <- submit operator do exerciseCmd tranche1Cid Structuring.InstructInvestorSettlement with confirmationCids = confirmationCids1; price; dateOfSettlement; dateOfTrade
   dvpCids32 <- submit operator do exerciseCmd tranche2Cid Structuring.InstructInvestorSettlement with confirmationCids = confirmationCids2; price; dateOfSettlement; dateOfTrade
-  dvpCids33 <- submit operator do exerciseCmd tranche3Cid Structuring.InstructInvestorSettlement with confirmationCids = confirmationCids3; price; dateOfSettlement; dateOfTrade
+  dvpCids33 <- submit operator do exerciseCmd tranche3Cid Structuring.InstructInvestorSettlementHtlc with confirmationCids = confirmationCids3; price; dateOfSettlement; dateOfTrade; hashlock; expiry
```
Comment (GeorgSchneider) [inline daml/Tests/Distribution/Syndication/Issuance.daml:126]: For investor settlement we need to think who will provide the hashlock. it'll be different for each investor, so maybe it's the investor providing it on the confirmation

File: daml/Tests/Distribution/Syndication/Issuance.daml (hunk 1 above)
Comment (judywu-da) [reply daml/Tests/Distribution/Syndication/Issuance.daml:115]: yes, you are right, I was such a lazy person, updated.

File: daml/Tests/Distribution/Syndication/Issuance.daml (hunk 2 above)
Comment (judywu-da) [reply daml/Tests/Distribution/Syndication/Issuance.daml:126]: ah right, each investor should all provide a different hashlock and hold their own secrets.
I will add hashlock into confirmation as commented.
Thanks for the catch.

Comment (judywu-da) [reply daml/Tests/Distribution/Syndication/Issuance.daml:126]: I put an Optional hashlock into Confirmation, probably not the best way.
I was originally thinking of making SettlementMode something like
`SettlementMode = Fop | Dvp | Htlc hashlock: Text` but it doesn't fit well in the current Tranche template and made it harder to work with.

**Review by dorrit-du-da — APPROVED** [review]

**Pull request merged**